File: sdks/nodejs/src/bridge.js
'use strict';

const EventEmitter = require('events');
const WebSocket = require('ws');
const os = require('os');
const { createMessage, parseMessage, authApiKey, authDeviceToken, authUserPassword } = require('./protocol');

/**
 * CDAPBridge — connects a Node.js process to Yomie CDAP Gateway.
 *
 * @fires CDAPBridge#connected
 * @fires CDAPBridge#disconnected
 * @fires CDAPBridge#registered
 * @fires CDAPBridge#command
 * @fires CDAPBridge#error
 * @fires CDAPBridge#message
 */
class CDAPBridge extends EventEmitter {
  /**
   * @param {object} options
   * @param {string} options.server - WebSocket URL, e.g. ws://host:21122/cdap
   * @param {string} [options.authMethod='api_key'] - api_key | device_token | user_password
   * @param {string} [options.apiKey]
   * @param {string} [options.deviceToken]
   * @param {string} [options.username]
   * @param {string} [options.password]
   * @param {string} [options.deviceId]
   * @param {string} [options.deviceName]
   * @param {string} [options.deviceType='bridge']
   * @param {string} [options.bridgeName]
   * @param {string} [options.bridgeVersion='1.0.0']
   * @param {number} [options.heartbeatSec=15]
   * @param {number} [options.reconnectSec=5]
   * @param {number} [options.maxReconnect=120]
   */
  constructor(options) {
    super();
    this.server = options.server;
    this.authMethod = options.authMethod || 'api_key';
    this.apiKey = options.apiKey || '';
    this.deviceToken = options.deviceToken || '';
    this.username = options.username || '';
    this.password = options.password || '';
    this.deviceId = options.deviceId || '';
    this.deviceName = options.deviceName || os.hostname();
    this.deviceType = options.deviceType || 'bridge';
    this.bridgeName = options.bridgeName || '';
    this.bridgeVersion = options.bridgeVersion || '1.0.0';
    this.heartbeatSec = options.heartbeatSec || 15;
    this.reconnectSec = options.reconnectSec || 5;
    this.maxReconnect = options.maxReconnect || 120;

    /** @type {Map<string, import('./widgets').Widget>} */
    this._widgets = new Map();
    /** @type {Map<string, any>} */
    this._state = new Map();
    /** @type {Map<string, Function>} */
    this._handlers = new Map();

    this._ws = null;
    this._heartbeatTimer = null;
    this._token = null;
    this._running = false;
    this._backoff = this.reconnectSec;
  }

  // ── Widget registration ──────────────────────────────────────────

  /**
   * Add a widget to the manifest. Must be called before connect().
   * @param {import('./widgets').Widget} widget
   * @returns {this}
   */
  addWidget(widget) {
    this._widgets.set(widget.id, widget);
    return this;
  }

  /**
   * Register a command handler for a specific widget.
   * @param {string} widgetId
   * @param {function(object): any} handler - receives command payload, may return a value
   * @returns {this}
   */
  onCommand(widgetId, handler) {
    this._handlers.set(widgetId, handler);
    return this;
  }

  // ── State updates ────────────────────────────────────────────────

  /**
   * Update a single widget value and push to server.
   * @param {string} widgetId
   * @param {any} value
   */
  updateState(widgetId, value) {
    this._state.set(widgetId, value);
    this._send('state_update', { widget_id: widgetId, value });
  }

  /**
   * Bulk-update multiple widget values.
   * @param {Object<string, any>} updates - { widgetId: value, ... }
   */
  bulkUpdate(updates) {
    const values = {};
    for (const [k, v] of Object.entries(updates)) {
      this._state.set(k, v);
      values[k] = v;
    }
    this._send('state_update', { values });
  }

  // ── Alerts & logging ─────────────────────────────────────────────

  /**
   * Fire an alert.
   * @param {string} alertId
   * @param {string} severity - critical | warning | info
   * @param {string} message
   * @param {object} [details]
   */
  fireAlert(alertId, severity, message, details) {
    this._send('alert', { alert_id: alertId, severity, message, details: details || {} });
  }

  /**
   * Resolve a previously fired alert.
   * @param {string} alertId
   */
  resolveAlert(alertId) {
    this._send('alert_resolve', { alert_id: alertId });
  }

  /**
   * Send a log entry.
   * @param {string} level - debug | info | warn | error
   * @param {string} message
   * @param {object} [data]
   */
  sendLog(level, message, data) {
    this._send('log', { level, message, data: data || {} });
  }

  // ── Connection lifecycle ─────────────────────────────────────────

  /**
   * Start the bridge — connects, authenticates, registers, and maintains heartbeat.
   * Reconnects automatically on disconnection.
   * @returns {Promise<void>} resolves when stop() is called
   */
  async run() {
    this._running = true;
    this._backoff = this.reconnectSec;

    while (this._running) {
      try {
        await this._connectOnce();
        // If we get here, connection was cleanly closed
      } catch (err) {
        if (!this._running) break;
        this.emit('error', err);
      }

      if (!this._running) break;

      // Exponential backoff with jitter
      const jitter = Math.floor(Math.random() * (this._backoff / 4));
      const delay = this._backoff + jitter;
      await this._sleep(delay * 1000);
      this._backoff = Math.min(this._backoff * 2, this.maxReconnect);
    }
  }

  /**
   * Stop the bridge gracefully.
   */
  stop() {
    this._running = false;
    this._clearHeartbeat();
    if (this._ws && this._ws.readyState === WebSocket.OPEN) {
      this._ws.close(1000, 'bridge stopping');
    }
    this._ws = null;
  }

  // ── Internal ─────────────────────────────────────────────────────

  /** @private */
  async _connectOnce() {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(this.server, 'cdap', {
        headers: { 'User-Agent': `yomie-cdap-nodejs/${this.bridgeVersion}` },
        handshakeTimeout: 10000,
      });

      let authenticated = false;

      ws.on('open', () => {
        this._ws = ws;
        this._backoff = this.reconnectSec; // reset backoff on success
        this.emit('connected');
        this._authenticate();
      });

      ws.on('message', (raw) => {
        let msg;
        try {
          msg = parseMessage(raw);
        } catch {
          return;
        }

        this.emit('message', msg);

        switch (msg.type) {
          case 'auth_result':
            if (msg.payload.success) {
              authenticated = true;
              this._token = msg.payload.token || this._token;
              if (msg.payload.device_id) this.deviceId = msg.payload.device_id;
              this._register();
            } else {
              ws.close(4001, 'auth failed');
              reject(new Error(`Auth failed: ${msg.payload.error || 'unknown'}`));
            }
            break;

          case 'registered':
            this.emit('registered', msg.payload);
            this._startHeartbeat();
            break;

          case 'command':
            this._handleCommand(msg);
            break;

          case 'heartbeat_ack':
            // noop
            break;

          case 'error':
            this.emit('error', new Error(msg.payload.message || 'server error'));
            break;

          default:
            // Forward unrecognized message types via event
            this.emit(msg.type, msg.payload);
            break;
        }
      });

      ws.on('close', (code, reason) => {
        this._clearHeartbeat();
        this._ws = null;
        this.emit('disconnected', { code, reason: reason.toString() });
        if (authenticated) {
          resolve(); // normal reconnect
        } else if (code !== 4001) {
          resolve();
        }
      });

      ws.on('error', (err) => {
        // error fires before close, let close handler resolve/reject
        if (!authenticated && !ws._closeCalled) {
          reject(err);
        }
      });
    });
  }

  /** @private */
  _authenticate() {
    let payload;
    switch (this.authMethod) {
      case 'device_token':
        payload = authDeviceToken(this.deviceToken, this.deviceId, this.bridgeVersion);
        break;
      case 'user_password':
        payload = authUserPassword(this.username, this.password, this.deviceId, this.bridgeVersion);
        break;
      default:
        payload = authApiKey(this.apiKey, this.deviceId, this.bridgeVersion);
        break;
    }
    this._send('auth', payload);
  }

  /** @private */
  _register() {
    const widgets = [];
    for (const w of this._widgets.values()) {
      widgets.push(typeof w.toJSON === 'function' ? w.toJSON() : w);
    }

    const capabilities = ['telemetry', 'commands', 'alerts', 'logs'];
    const manifest = {
      manifest_version: '1.0',
      device: {
        id: this.deviceId,
        name: this.deviceName,
        type: this.deviceType,
        firmware_version: this.bridgeVersion,
        hostname: os.hostname(),
        os: `${os.type()} ${os.release()}`,
        platform: os.platform(),
      },
      capabilities,
      widgets,
      heartbeat_interval: this.heartbeatSec,
    };

    if (this.bridgeName) {
      manifest.bridge = { name: this.bridgeName, version: this.bridgeVersion };
    }

    this._send('register', manifest);
  }

  /** @private */
  _startHeartbeat() {
    this._clearHeartbeat();
    this._heartbeatTimer = setInterval(() => {
      const widgetValues = {};
      for (const [k, v] of this._state) {
        widgetValues[k] = v;
      }
      this._send('heartbeat', {
        uptime: Math.floor(process.uptime()),
        widget_values: widgetValues,
      });
    }, this.heartbeatSec * 1000);
  }

  /** @private */
  _clearHeartbeat() {
    if (this._heartbeatTimer) {
      clearInterval(this._heartbeatTimer);
      this._heartbeatTimer = null;
    }
  }

  /** @private */
  _handleCommand(msg) {
    const { widget_id, action, value, params } = msg.payload;
    const handler = this._handlers.get(widget_id);

    this.emit('command', msg.payload);

    if (!handler) {
      this._send('command_response', {
        id: msg.id,
        widget_id,
        success: false,
        error: `No handler for widget ${widget_id}`,
      });
      return;
    }

    const start = Date.now();
    try {
      const result = handler({ widget_id, action, value, params });

      // Handle both sync and async handlers
      if (result && typeof result.then === 'function') {
        result
          .then((val) => {
            this._sendCommandResponse(msg.id, widget_id, true, val, Date.now() - start);
            if (val !== undefined) {
              this._state.set(widget_id, val);
            }
          })
          .catch((err) => {
            this._sendCommandResponse(msg.id, widget_id, false, null, Date.now() - start, err.message);
          });
      } else {
        this._sendCommandResponse(msg.id, widget_id, true, result, Date.now() - start);
        if (result !== undefined) {
          this._state.set(widget_id, result);
        }
      }
    } catch (err) {
      this._sendCommandResponse(msg.id, widget_id, false, null, Date.now() - start, err.message);
    }
  }

  /** @private */
  _sendCommandResponse(id, widgetId, success, value, durationMs, error) {
    const payload = { id, widget_id: widgetId, success, duration_ms: durationMs };
    if (value !== undefined && value !== null) payload.value = value;
    if (error) payload.error = error;
    this._send('command_response', payload);
  }

  /** @private */
  _send(type, payload) {
    if (this._ws && this._ws.readyState === WebSocket.OPEN) {
      this._ws.send(createMessage(type, payload));
    }
  }

  /** @private */
  _sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

module.exports = { CDAPBridge };

File: sdks/python/yomie_cdap/bridge.py
"""CDAPBridge — main class for building CDAP device bridges in Python.

Usage example:

    from betterdesk_cdap import CDAPBridge, gauge, toggle

    bridge = CDAPBridge(
        server="ws://192.168.0.110:21122/cdap",
        auth_method="api_key",
        api_key="your-key",
        device_name="Temperature Sensor",
        device_type="iot",
    )

    temp = bridge.add_widget(gauge("temp", "Temperature", unit="°C", max_val=50))
    heater = bridge.add_widget(toggle("heater", "Heater"))

    @bridge.on_command("heater")
    async def handle_heater(action, value, **kw):
        GPIO.output(HEATER_PIN, value)
        return {"status": "ok"}

    bridge.run()
"""

from __future__ import annotations

import asyncio
import json
import logging
import random
import time
from typing import Any, Callable, Coroutine

import websockets
import websockets.exceptions

from betterdesk_cdap.protocol import Message, auth_payload_api_key, auth_payload_device_token, auth_payload_user_password
from betterdesk_cdap.widgets import Widget

logger = logging.getLogger("betterdesk_cdap")


class CDAPBridge:
    """Python bridge that connects to a Yomie CDAP gateway."""

    def __init__(
        self,
        server: str,
        *,
        auth_method: str = "api_key",
        api_key: str = "",
        device_token: str = "",
        username: str = "",
        password: str = "",
        device_id: str = "",
        device_name: str = "Python Bridge",
        device_type: str = "iot",
        vendor: str = "",
        model: str = "",
        firmware: str = "1.0.0",
        tags: list[str] | None = None,
        capabilities: list[str] | None = None,
        heartbeat_sec: int = 15,
        reconnect_sec: int = 5,
        max_reconnect: int = 300,
        # Bridge metadata (optional)
        bridge_name: str = "",
        bridge_protocol: str = "",
        target_host: str = "",
        target_port: int = 0,
    ):
        self.server = server
        self.auth_method = auth_method
        self.api_key = api_key
        self.device_token = device_token
        self.username = username
        self.password = password
        self.device_id = device_id
        self.device_name = device_name
        self.device_type = device_type
        self.vendor = vendor
        self.model = model
        self.firmware = firmware
        self.tags = tags or []
        self.capabilities = capabilities or ["telemetry", "commands"]
        self.heartbeat_sec = max(5, min(300, heartbeat_sec))
        self.reconnect_sec = max(1, reconnect_sec)
        self.max_reconnect = max(reconnect_sec, max_reconnect)

        self.bridge_name = bridge_name
        self.bridge_protocol = bridge_protocol
        self.target_host = target_host
        self.target_port = target_port

        # Internal state
        self._ws: websockets.WebSocketClientProtocol | None = None
        self._widgets: list[Widget] = []
        self._widget_values: dict[str, Any] = {}
        self._command_handlers: dict[str, Callable] = {}
        self._global_command_handler: Callable | None = None
        self._token: str = ""
        self._assigned_device_id: str = ""
        self._role: str = ""
        self._running = False
        self._connected = False

    # ── Widget Registration ───────────────────────────────────────────

    def add_widget(self, widget: Widget) -> Widget:
        """Register a widget for the device manifest."""
        self._widgets.append(widget)
        if widget.value is not None:
            self._widget_values[widget.id] = widget.value
        return widget

    def on_command(self, widget_id: str | None = None):
        """Decorator to register a command handler for a widget (or all widgets)."""
        def decorator(fn: Callable[..., Coroutine]):
            if widget_id:
                self._command_handlers[widget_id] = fn
            else:
                self._global_command_handler = fn
            return fn
        return decorator

    # ── State Updates ─────────────────────────────────────────────────

    async def update_state(self, widget_id: str, value: Any) -> None:
        """Push a single widget state update to the server."""
        self._widget_values[widget_id] = value
        await self._send("state_update", {
            "widget_id": widget_id,
            "value": value,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        })

    async def bulk_update(self, updates: dict[str, Any]) -> None:
        """Push multiple widget state updates."""
        for wid, val in updates.items():
            self._widget_values[wid] = val
        await self._send("bulk_update", {
            "updates": [
                {"widget_id": wid, "value": val}
                for wid, val in updates.items()
            ],
        })

    # ── Alerts ────────────────────────────────────────────────────────

    async def fire_alert(
        self,
        alert_id: str,
        severity: str = "warning",
        message: str = "",
        data: Any = None,
    ) -> None:
        """Fire an alert to the server."""
        payload: dict[str, Any] = {
            "alert_id": alert_id,
            "severity": severity,
            "message": message,
        }
        if data is not None:
            payload["data"] = data
        await self._send("event", {"event_type": "alert_fire", "data": payload})

    async def resolve_alert(self, alert_id: str) -> None:
        """Resolve a previously fired alert."""
        await self._send("event", {
            "event_type": "alert_resolve",
            "data": {"alert_id": alert_id},
        })

    # ── Logging ───────────────────────────────────────────────────────

    async def send_log(
        self, level: str, message: str, context: Any = None
    ) -> None:
        """Send a log entry to the server."""
        payload: dict[str, Any] = {"level": level, "message": message}
        if context is not None:
            payload["context"] = context
        await self._send("log", payload)

    # ── Lifecycle ─────────────────────────────────────────────────────

    def run(self) -> None:
        """Blocking entry point — connect and run the message loop."""
        asyncio.run(self._run_forever())

    async def _run_forever(self) -> None:
        self._running = True
        delay = self.reconnect_sec

        while self._running:
            try:
                await self._connect_once()
                delay = self.reconnect_sec  # reset on success
            except Exception as exc:
                logger.warning("Connection lost: %s", exc)

            if not self._running:
                break

            jitter = random.uniform(0, delay * 0.25)
            wait = delay + jitter
            logger.info("Reconnecting in %.1fs...", wait)
            await asyncio.sleep(wait)
            delay = min(delay * 2, self.max_reconnect)

    async def _connect_once(self) -> None:
        logger.info("Connecting to %s...", self.server)
        async with websockets.connect(
            self.server,
            max_size=4 * 1024 * 1024,
            open_timeout=30,
        ) as ws:
            self._ws = ws
            await self._authenticate()
            await self._register()
            self._connected = True
            logger.info(
                "Connected as %r (device_id=%s, role=%s)",
                self.device_name,
                self._assigned_device_id,
                self._role,
            )

            heartbeat_task = asyncio.create_task(self._heartbeat_loop())
            try:
                await self._message_loop()
            finally:
                heartbeat_task.cancel()
                self._connected = False
                self._ws = None

    def stop(self) -> None:
        """Signal the bridge to stop."""
        self._running = False

    # ── Authentication ────────────────────────────────────────────────

    async def _authenticate(self) -> None:
        if self.auth_method == "api_key":
            payload = auth_payload_api_key(self.api_key, self.device_id, self.firmware)
        elif self.auth_method == "device_token":
            payload = auth_payload_device_token(self.device_token, self.device_id, self.firmware)
        elif self.auth_method == "user_password":
            payload = auth_payload_user_password(
                self.username, self.password, self.device_id, self.firmware
            )
        else:
            raise ValueError(f"Unknown auth method: {self.auth_method}")

        await self._send("auth", payload)
        msg = await self._recv()
        if msg.type != "auth_result":
            raise RuntimeError(f"Expected auth_result, got {msg.type}")
        if not msg.payload.get("success"):
            raise RuntimeError(f"Auth failed: {msg.payload.get('error', 'unknown')}")

        self._token = msg.payload.get("token", "")
        self._assigned_device_id = msg.payload.get("device_id", "")
        self._role = msg.payload.get("role", "")

    # ── Registration ──────────────────────────────────────────────────

    async def _register(self) -> None:
        manifest = self._build_manifest()
        await self._send("register", {"manifest": manifest})

    def _build_manifest(self) -> dict[str, Any]:
        device: dict[str, Any] = {
            "name": self.device_name,
            "type": self.device_type,
        }
        if self.vendor:
            device["vendor"] = self.vendor
        if self.model:
            device["model"] = self.model
        if self.firmware:
            device["firmware"] = self.firmware
        if self.tags:
            device["tags"] = self.tags

        m: dict[str, Any] = {
            "manifest_version": "1.0",
            "device": device,
            "capabilities": self.capabilities,
            "heartbeat_interval": self.heartbeat_sec,
            "widgets": [w.to_dict() for w in self._widgets],
        }

        if self.bridge_name:
            m["bridge"] = {
                "name": self.bridge_name,
                "version": self.firmware,
                "protocol": self.bridge_protocol,
                "target_host": self.target_host,
                "target_port": self.target_port,
            }

        return m

    # ── Heartbeat ─────────────────────────────────────────────────────

    async def _heartbeat_loop(self) -> None:
        while self._connected:
            try:
                payload: dict[str, Any] = {}
                if self._widget_values:
                    payload["widget_values"] = dict(self._widget_values)
                await self._send("heartbeat", payload)
            except Exception:
                return
            await asyncio.sleep(self.heartbeat_sec)

    # ── Message Loop ──────────────────────────────────────────────────

    async def _message_loop(self) -> None:
        while True:
            msg = await self._recv()
            await self._dispatch(msg)

    async def _dispatch(self, msg: Message) -> None:
        if msg.type == "command":
            await self._handle_command(msg.payload)
        elif msg.type == "ping":
            pass  # heartbeat ACK
        elif msg.type == "error":
            code = msg.payload.get("code", 0)
            error_msg = msg.payload.get("message", "")
            logger.error("Server error %d: %s", code, error_msg)
        elif msg.type == "registered":
            logger.debug("Registration confirmed: %s", msg.payload)
        else:
            logger.debug("Unhandled message type: %s", msg.type)

    async def _handle_command(self, payload: dict[str, Any]) -> None:
        command_id = payload.get("command_id", "")
        widget_id = payload.get("widget_id", "")
        action = payload.get("action", "")
        value = payload.get("value")

        start = time.monotonic()
        handler = self._command_handlers.get(widget_id, self._global_command_handler)

        resp: dict[str, Any] = {"command_id": command_id}
        if handler is None:
            resp["status"] = "error"
            resp["error_message"] = f"No handler for widget {widget_id}"
        else:
            try:
                result = await handler(
                    action=action,
                    value=value,
                    widget_id=widget_id,
                    operator=payload.get("operator", ""),
                    reason=payload.get("reason", ""),
                )
                resp["status"] = "ok"
                resp["result"] = result
                # Auto-update widget value on set
                if action == "set" and value is not None:
                    self._widget_values[widget_id] = value
            except Exception as exc:
                resp["status"] = "error"
                resp["error_message"] = str(exc)

        resp["execution_time_ms"] = int((time.monotonic() - start) * 1000)
        await self._send("command_response", resp)

    # ── Wire I/O ──────────────────────────────────────────────────────

    async def _send(self, msg_type: str, payload: dict[str, Any]) -> None:
        if not self._ws:
            return
        msg = Message(type=msg_type, payload=payload)
        await self._ws.send(msg.to_json())

    async def _recv(self) -> Message:
        if not self._ws:
            raise RuntimeError("Not connected")
        raw = await self._ws.recv()
        return Message.from_json(raw)
